erigon-lib/state/domain.go

```diff
@@ -836,47 +874,75 @@ func (d *Domain) collate(ctx context.Context, step, txFrom, txTo uint64, roTx kv
 	}
 	comp := NewArchiveWriter(coll.valuesComp, d.compression)
 
-	keysCursor, err := roTx.CursorDupSort(d.keysTable)
-	if err != nil {
-		return Collation{}, fmt.Errorf("create %s keys cursor: %w", d.filenameBase, err)
-	}
-	defer keysCursor.Close()
+	stepBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(stepBytes, ^step)
 
-	var (
-		stepBytes = make([]byte, 8)
-		keySuffix = make([]byte, 256+8)
-		v         []byte
+	var valsCursor kv.Cursor
 
-		valsDup kv.CursorDupSort
-	)
-	binary.BigEndian.PutUint64(stepBytes, ^step)
-	valsDup, err = roTx.CursorDupSort(d.valsTable)
-	if err != nil {
-		return Collation{}, fmt.Errorf("create %s values cursorDupsort: %w", d.filenameBase, err)
+	if d.largeVals {
+		valsCursor, err = roTx.Cursor(d.valsTable)
+		if err != nil {
+			return Collation{}, fmt.Errorf("create %s values cursorDupsort: %w", d.filenameBase, err)
+		}
+	} else {
+		valsCursor, err = roTx.CursorDupSort(d.valsTable)
+		if err != nil {
+			return Collation{}, fmt.Errorf("create %s values cursorDupsort: %w", d.filenameBase, err)
+		}
 	}
-	defer valsDup.Close()
+	defer valsCursor.Close()
+
+	kvs := []struct {
+		k, v []byte
+	}{}
 
-	for k, stepInDB, err := keysCursor.First(); k != nil; k, stepInDB, err = keysCursor.Next() {
+	var stepInDB []byte
+	for k, v, err := valsCursor.First(); k != nil; {
 		if err != nil {
 			return coll, err
 		}
+
+		if d.largeVals {
+			stepInDB = k[len(k)-8:]
+		} else {
+			stepInDB = v[:8]
+		}
 		if !bytes.Equal(stepBytes, stepInDB) { // [txFrom; txTo)
+			k, v, err = valsCursor.Next()
 			continue
 		}
 
-		copy(keySuffix, k)
-		copy(keySuffix[len(k):], stepInDB)
-
-		v, err = roTx.GetOne(d.valsTable, keySuffix[:len(k)+8])
-		if err != nil {
-			return coll, fmt.Errorf("find last %s value for aggregation step k=[%x]: %w", d.filenameBase, k, err)
+		if d.largeVals {
+			kvs = append(kvs, struct {
+				k, v []byte
+			}{common.Copy(k[:len(k)-8]), common.Copy(v)})
+			k, v, err = valsCursor.Next()
+		} else {
+			if err = comp.AddWord(k); err != nil {
```

Comment:
i see AddWord called here and after loop i see `if err = comp.AddWord(kv.k); err != nil {`. why need 2 places for AddWord?

Reply:
see comment on the PR, it is because the order is not what you would expect as inverted_step gets in the way